cpp/oai_pmh_harvester.cc

```diff
@@ -170,85 +157,179 @@ bool ListRecords(const std::string &url, const unsigned time_limit_in_seconds_pe
-            const std::string control_number(control_number_prefix + StringUtil::Map(StringUtil::ToString(++counter, 10, 10), ' ', '0'));
+            const std::string control_number(control_number_prefix
+                                             + StringUtil::Map(StringUtil::ToString(record_number, 10, 10), ' ', '0'));
             record.insertField("001", control_number);
         }
 
-        marc_writer->write(record);
+        if (WriteIfNotDupe(record, dups_db, marc_writer))
+            counter++;
     }
+
+    LOG_INFO("Extracted " + std::to_string(counter) + " MARC record(s).");
 }
 
 
 } // unnamed namespace
 
 
 int Main(int argc, char **argv) {
-    std::unique_ptr<KeyValueDB> dups_db;
-    if (argc > 1 and std::strcmp(argv[1], "--skip-dups") == 0) {
-        dups_db = CreateOrOpenKeyValueDB();
+    if (argc > 1 and std::strcmp(argv[1], "--skip-dups") == 0)
         --argc, ++argv;
-    }
 
     bool ignore_ssl_certificates(false);
     if (argc > 1 and std::strcmp(argv[1], "--ignore-ssl-certificates") == 0) {
        ignore_ssl_certificates = true;
        --argc, ++argv;
     }
 
-    if (argc != 6 and argc != 7)
-        Usage();
+    if (argc != 7 and argc != 8)
+        ::Usage("[--skip-dups] [--ignore-ssl-certificates] base_url metadata_prefix [harvest_set_or_identifier] "
+                "control_number_prefix output_filename time_limit_per_request path_to_dups_database\n"
+                "If \"--skip-dups\" has been specified, records that we already encountered in the past won't\n"
+                "included in the output file.\n"
+                "\"harvest_set_or_identifier\" must start with \"set=\" or \"identifier=\".\n"
+                "\"control_number_prefix\" will be used if the received records have no control numbers\n"
+                "to autogenerate our own control numbers.  \"time_limit_per_request\" is in seconds. (Some\n"
+                "servers are very slow so we recommend at least 20 seconds!)\n");
 
     const std::string base_url(argv[1]);
+
     const std::string metadata_prefix(argv[2]);
-    const std::string harvest_set(argc == 7 ? argv[3] : "");
+    if (metadata_prefix != "marc" and metadata_prefix != "oai_dc")
+        LOG_ERROR("metadata_prefix must be \"marc\" or \"oai_dc\"!");
+
+    const std::string harvest_set_or_identifier(argc == 7 ? argv[3] : "");
+    if (not harvest_set_or_identifier.empty()
+        and (not StringUtil::StartsWith(harvest_set_or_identifier, "set=")
+             and not StringUtil::StartsWith(harvest_set_or_identifier, "identifier=")))
+        LOG_ERROR("harvest_set_or_identifier must start with set= or identifier=!");
+
     const std::string control_number_prefix(argc == 7 ? argv[3] : argv[2]);
     const std::string output_filename(argc == 7 ? argv[5] : argv[4]);
     const std::string time_limit_per_request_as_string(argc == 7 ? argv[6] : argv[5]);
+    const std::string path_to_dups_database(argc == 7 ? argv[7] : argv[6]);
```

Comment:
Is this really correct? `argv[7] ` will not exist if `argc == 7`

Reply:
No, that was wrong, but it is fixed now.